### ISSUE TITLE
Error `async operation inside` in r.subrequest with callback

### ISSUE THREAD

**Issue by levonet** [nginx/njs#287]
nginx throw errors after update version of njs to 0.3.8:
`2020/02/10 16:18:29 [error] 144#144: *6947506 async operation inside "js_log" variable handler, client: nda, server: nda, request: "POST /nda HTTP/1.1", host: "nda"`

`r.subrequest ` continues to work. In documentation:

> Since 0.3.8, if a `callback` is not provided, the `Promise` object that resolves to the subrequest response object is returned.

There should be no error if a function has a callback argument. I think so.

Here is a minimal case:
```js
function js_log(r) {
    var log = {
        // ...
    };
    var args = {
        method: 'POST',
        body: JSON.stringify(log)
    };
    r.subrequest('/_log', args, (res) => {});
    return '1';
}
```
in conf:
```nginx
js_set $x_js_log js_log;
server {
    location / {
        proxy_set_header X-Log $x_js_log;
        proxy_pass http://backend;
    }
    location = /_log {
        internal;
        return 200;
    }
}
```

**Comment by xeioex** [comment]
Hi @levonet ,

nginx does not support asynchronous operations (it is not related to `async` keyword in JavaScript) in its variable handlers. When nginx want the value of some variable (say `$x_js_log`) it finds it and calls its handler (js_log). Nginx cannot block here, and wants the result right away. r.subrequest() is asynchronous operation by it nature.

To avoid confusion, nginx-http-js module [checks](https://github.com/nginx/njs/blob/master/nginx/ngx_http_js_module.c#L678) whether pending operations was created in variable handler and returns error in such situation.

Yes, that check can be removed, in the sense that r.subrequest() can be scheduled from `js_set` handler, but there is no way to return r.subrequest() result as `js_set` result.

Can you give a good example where such behavior is desired?

**Comment by levonet** [comment]
I agree with you. But in my case proxy_pass and r.subrequest() work in parallel.
r.subrequest() called to send part of requestBody to storage. And `js_log` not wait for a response.
`js_log` always return `'1'`.

**Comment by xeioex** [comment]
@levonet 

BTW, take a look at http://nginx.org/en/docs/http/ngx_http_mirror_module.html Maybe this is what you actually need?

**Comment by levonet** [comment]
@xeioex Thanks!
This works fine with `js_content`.

**Comment by levonet** [comment]
I didn't remember, but maybe I refused mirror after the first overload tests. Now I use queue before storage.
In other cases, maybe, need custom switch off of the promise check in `js_set`.

**Comment by xeioex** [comment]
In 0.3.8 background subrequests were removed, which is almost the same as mirror directive. Before 0.3.8 absense of a callback meant `background subrequest`, after 0.3.8 promise object which installs its own callback.

We can consider adding background flag, to `options` argument to enforce background subrequests.

**Comment by xeioex** [comment]
@levonet 

You may try the [patch](https://gist.github.com/xeioex/ae0eac8c796a658bf367f3ab0ce84055).

**Comment by xeioex** [comment]
@levonet 
Was committed in https://github.com/nginx/njs/commit/889c17e1bae57021605af7831ab15935330b0b6d

The flag was changed to "detached". Will be released within 0.3.9 (tomorrow, 03.03.2020).
